### PR TITLE
Improve the documentation for Vec::drain

### LINKED ISSUE #73844: drain() doc of vector is unclear
I wanted to move out a number of front element from a vector, like `my_vec.drain(..x)` but I wonder if this is the right way to do this cause doc is unclear, note 1 and note 2 say the opposite each other:

> Creates a draining iterator that removes the specified range in the vector and yields the removed items.
>
> Note 1: The element range is removed even if the iterator is only partially consumed or not consumed at all.
>
> Note 2: It is unspecified how many elements are removed from the vector if the Drain value is leaked.

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain

This issue has been assigned to @poliorcetics via [this comment](https://github.com/rust-lang/rust/issues/73844#issuecomment-662657778).

### PR DESCRIPTION
**Pull request by poliorcetics**
Fixes #73844.

@rusbot modify labels: A-collections, C-enhancement, T-doc, T-libs

### CODE REVIEW

File: src/liballoc/vec.rs [hunk 1]
```diff
@@ -1297,6 +1298,8 @@ impl<T> Vec<T> {
     /// v.drain(..);
     /// assert_eq!(v, &[]);
     /// ```
+    ///
+    /// [`mem::forget`]: mem::forget
```
Comment (jyn514) [inline src/liballoc/vec.rs:1302]: You don't need this link, rustdoc will automatically strip the backticks.

File: src/liballoc/vec.rs [hunk 2]
```diff
@@ -1274,11 +1274,12 @@ impl<T> Vec<T> {
...
-    /// Note 1: The element range is removed even if the iterator is only
-    /// partially consumed or not consumed at all.
+    /// The element range is removed even if the iterator is only partially
+    /// consumed or not consumed at all.
     ///
-    /// Note 2: It is unspecified how many elements are removed from the vector
-    /// if the `Drain` value is leaked.
+    /// Note: Be aware that if the iterator is leaked (eg: [`mem::forget`]), it
+    /// cancels this property so it is unspecified how many elements are removed
+    /// from the vector in this case.
```
Comment (jyn514) [inline src/liballoc/vec.rs:1282]: How about this:

> When the iterator is dropped, all elements in the range are removed from the vector, even if the iterator was not fully consumed.
> If the iterator is not dropped, it is unspecified how many elements are removed.

Comment (poliorcetics) [reply src/liballoc/vec.rs]: Done !

**Pull request merged**